### PR TITLE
Java: Add command `Zrangestore`. (Sorted Set Group)

### PR DESCRIPTION
**Pull request by SanHalacogluImproving**

### CODE REVIEW

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 1]
```diff
@@ -413,6 +413,71 @@ CompletableFuture<Map<String, Double>> zrangeWithScores(
...
+     * @param destination The key for the destination sorted set.
+     * @param source The key of the source sorted set.
+     * @param rangeQuery The range query object representing the type of range query to perform.<br>
+     *     <ul>
+     *       <li>For range queries by index (rank), use {@link RangeByIndex}.
+     *       <li>For range queries by lexicographical order, use {@link RangeByLex}.
+     *       <li>For range queries by score, use {@link RangeByScore}.
+     *     </ul>
+     *
+     * @param reverse If true, reverses the sorted set, with index 0 as the element with the highest
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:433]: wrap `true` and 0 with `code`

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 2]
```diff
@@ -413,6 +413,71 @@ CompletableFuture<Map<String, Double>> zrangeWithScores(
...
+     *     </ul>
+     *
+     * @return The number of elements in the resulting sorted set.
+     * @example
+     *     <pre>{@code
+     * RangeByIndex query1 = new RangeByIndex(0, -1); // Query for all members.
+     * Long payload1 = client.zrange("destinationKey","mySortedSet",query1).get();
+     * assert payload1 == 7L;
+     *
+     * RangeByScore query2 = new RangeByScore(InfScoreBound.NEGATIVE_INFINITY, new ScoreBoundary(3)); // Query for members with scores within the range of negative infinity to 3.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:474]:
```suggestion
     * RangeByScore query2 = new RangeByScore(NEGATIVE_INFINITY, new ScoreBoundary(3)); // Query for members with scores within the range of negative infinity to 3.
```

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 3]
```diff
@@ -413,6 +413,71 @@ CompletableFuture<Map<String, Double>> zrangeWithScores(
...
+     */
+    CompletableFuture<Long> zrangestore(
+            String destination, String source, RangeQuery rangeQuery, boolean reverse);
+
+    /**
+     * Stores a specified range of elements from the sorted set at <code>source</code>, into a new
+     * sorted set at <code>destination</code>. If <code>destination</code> doesn't exist, a new sorted
+     * set is created; if it exists, it's overwritten.<br>
+     * <code>ZRANGESTORE</code> can perform different types of range queries: by index (rank), by the
+     * score, or by lexicographical order.<br>
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:455]: Should we add cross@links between these commands?

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 4]
```diff
@@ -1471,6 +1473,62 @@ public T zrankWithScore(@NonNull String key, @NonNull String member) {
...
+     * @param destination The key for the destination sorted set.
+     * @param source The key of the source sorted set.
+     * @param rangeQuery The range query object representing the type of range query to perform.<br>
+     *     <ul>
+     *       <li>For range queries by index (rank), use {@link RangeOptions.RangeByIndex}.
+     *       <li>For range queries by lexicographical order, use {@link RangeOptions.RangeByLex}.
+     *       <li>For range queries by score, use {@link RangeOptions.RangeByScore}.
+     *     </ul>
+     *
+     * @param reverse If true, reverses the sorted set, with index 0 as the element with the highest
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1493]: same

File: java/client/src/main/java/glide/api/models/commands/RangeOptions.java [hunk 5]
```diff
@@ -292,7 +293,23 @@ public RangeByScore(
 
     public static String[] createZrangeArgs(
             String key, RangeQuery rangeQuery, boolean reverse, boolean withScores) {
-        String[] arguments = new String[] {key, rangeQuery.getStart(), rangeQuery.getEnd()};
+        String[] arguments =
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/RangeOptions.java:296]: make those instance functions (not static)

File: java/client/src/test/java/glide/api/models/TransactionTests.java [hunk 6]
```diff
@@ -448,6 +449,27 @@ public void transaction_builds_protobuf_request(BaseTransaction<?> transaction)
                                 .addArgs(WITH_SCORE_REDIS_API)
                                 .build()));
 
+        transaction.zrangestore(
```
Comment (Yury-Fridlyand) [inline java/client/src/test/java/glide/api/models/TransactionTests.java:452]: add a test for function without reverse argument for coverage

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 7]
```diff
@@ -1275,6 +1275,175 @@ public void zrank(BaseClient client) {
...
+
+        // Non-existent range as start > stop.
+        assertEquals(0, client.zrangestore(destination, source, new RangeByIndex(3, 1)).get());
+        assertEquals(Map.of(), client.zrangeWithScores(destination, new RangeByIndex(0, -1)).get());
+
+        // Non-existing source.
+        assertEquals(0, client.zrangestore(destination, key, new RangeByIndex(0, -1)).get());
+        assertEquals(Map.of(), client.zrangeWithScores(destination, new RangeByIndex(0, -1)).get());
+
+        // Key exists, but it is not a set
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1308]: What if destination exists? add a test for all methods

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 8]
```diff
@@ -1275,6 +1275,175 @@ public void zrank(BaseClient client) {
...
+                Map.of("one", 1.0, "two", 2.0, "three", 3.0),
+                client.zrangeWithScores(destination, new RangeByIndex(0, -1)).get());
+
+        // Range from rank 0 to 1. In descending order of scores.
+        assertEquals(2, client.zrangestore(destination, source, new RangeByIndex(0, 1), true).get());
+        assertEquals(
+                Map.of("three", 3.0, "two", 2.0),
+                client.zrangeWithScores(destination, new RangeByIndex(0, -1)).get());
+
+        // Non-existent range as start > stop.
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1300]:
```suggestion
        // Incorrect range as start > stop.
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 9]
```diff
@@ -1275,6 +1275,175 @@ public void zrank(BaseClient client) {
...
+        //  Range from positive to negative infinity with rev set to true.Limited to ranks 1 to 2.
+        query =
+                new RangeByScore(
+                        InfScoreBound.POSITIVE_INFINITY, InfScoreBound.NEGATIVE_INFINITY, new Limit(1, 2));
+        assertEquals(2, client.zrangestore(destination, source, query, true).get());
+        assertEquals(
+                Map.of("two", 2.0, "one", 1.0),
+                client.zrangeWithScores(destination, new RangeByIndex(0, -1)).get());
+
+        // Non-existent range as start > stop.
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1360]:
```suggestion
        // Incorrect range as start > stop.
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 10]
```diff
@@ -1275,6 +1275,175 @@ public void zrank(BaseClient client) {
...
+        assertEquals(OK, client.set(key, "value").get());
+        ExecutionException executionException =
+                assertThrows(
+                        ExecutionException.class,
+                        () ->
+                                client
+                                        .zrangestore(
+                                                destination,
+                                                key,
+                                                new RangeByScore(new ScoreBoundary(0), new ScoreBoundary(3)))
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1380]: you can use existing query too
```suggestion
                                                query
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 11]
```diff
@@ -1275,6 +1275,175 @@ public void zrank(BaseClient client) {
...
+        assertEquals(OK, client.set(key, "value").get());
+        ExecutionException executionException =
+                assertThrows(
+                        ExecutionException.class,
+                        () ->
+                                client
+                                        .zrangestore(
+                                                destination,
+                                                key,
+                                                new RangeByLex(new LexBoundary("a"), new LexBoundary("c")))
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1442]: same (optional)
```suggestion
                                                query
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 12]
```diff
@@ -1275,6 +1275,175 @@ public void zrank(BaseClient client) {
...
+        // Range from negative to positive infinity. Limited to ranks 1 to 2.
+        query =
+                new RangeByLex(
+                        InfLexBound.NEGATIVE_INFINITY, InfLexBound.POSITIVE_INFINITY, new Limit(1, 2));
+        assertEquals(2, client.zrangestore(destination, source, query).get());
+        assertEquals(
+                Map.of("b", 2.0, "c", 3.0),
+                client.zrangeWithScores(destination, new RangeByIndex(0, -1)).get());
+
+        //  Range from positive to negative infinity with rev set to true.Limited to ranks 1 to 2.
```
Comment (jonathanl-bq) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1418]:
```suggestion
        // Range from positive to negative infinity with rev set to true. Limited to ranks 1 to 2.
```

**Review by jonathanl-bq — APPROVED** [review]
Conditional approval if you address Yury's comments.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java (hunk 3 above)
Comment (SanHalacogluImproving) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:455]: between what commands?

File: java/client/src/main/java/glide/api/models/commands/RangeOptions.java (hunk 5 above)
Comment (SanHalacogluImproving) [reply java/client/src/main/java/glide/api/models/commands/RangeOptions.java:296]: I would then have to initialize RangeOptions which I don't want to do.

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 7 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:1308]: Overwrites.

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 10 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:1380]: Variable used in lambda expression should be final or effectively final (you get this error)

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 11 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:1442]: Variable used in lambda expression should be final or effectively final (you get this error)

File: java/client/src/main/java/glide/api/models/commands/RangeOptions.java [hunk 13]
```diff
@@ -292,7 +293,23 @@ public RangeByScore(
...
+                concatenateArrays(new String[] {key}, createZrangeBaseArgs(rangeQuery, reverse));
+        if (withScores) {
+            arguments = concatenateArrays(arguments, new String[] {WITH_SCORES_REDIS_API});
+        }
+
+        return arguments;
+    }
+
+    public static String[] createZrangeStoreArgs(
```
[the remark this answers is not in the document]
Comment (SanHalacogluImproving) [reply java/client/src/main/java/glide/api/models/commands/RangeOptions.java:305]: The command name is zrange. What do you prefer?

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java [hunk 14]
```diff
@@ -413,6 +413,71 @@ CompletableFuture<Map<String, Double>> zrangeWithScores(
...
+     * @param reverse If true, reverses the sorted set, with index 0 as the element with the highest
+     *     score.
+     * @return The number of elements in the resulting sorted set.
+     * @example
+     *     <pre>{@code
+     * RangeByIndex query1 = new RangeByIndex(0, -1); // Query for all members.
+     * Long payload1 = client.zrange("destinationKey","mySortedSet",query1, true).get();
+     * assert payload1 == 7L;
+     *
+     * RangeByScore query2 = new RangeByScore(InfScoreBound.NEGATIVE_INFINITY, new ScoreBoundary(3)); // Query for members with scores within the range of negative infinity to 3.
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:442]: in javadoc you can do anything

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:455]: commands -> functions
zrangescore and zrangescore

File: java/client/src/main/java/glide/api/models/commands/RangeOptions.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/commands/RangeOptions.java:305]: `createZRangeStore...`
`.......^`

**Review by Yury-Fridlyand — APPROVED** [review]

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (SanHalacogluImproving) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:442]: Yeah but its confusing why not keep it like this.

File: java/client/src/test/java/glide/api/models/TransactionTests.java [hunk 15]
```diff
@@ -448,6 +450,38 @@ public void transaction_builds_protobuf_request(BaseTransaction<?> transaction)
...
+                "destination",
+                "source",
+                new RangeByScore(
+                        InfScoreBound.NEGATIVE_INFINITY, new ScoreBoundary(3, false), new Limit(1, 2)),
+                true);
+        results.add(
+                Pair.of(
+                        ZRangeStore,
+                        ArgsArray.newBuilder()
```
[the remark this answers is not in the document]
Comment (SanHalacogluImproving) [reply java/client/src/test/java/glide/api/models/TransactionTests.java:462]: Will do this on upstream when rebasing.

File: java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java (hunk 3 above)
Comment (SanHalacogluImproving) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:455]: I don't see any crosslinking in other commands with options etc.

Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java:455]: https://github.com/Bit-Quill/glide-for-redis/blob/adde68c835676f73a76c7833fb3422ceb0cc098e/java/client/src/main/java/glide/api/commands/SortedSetBaseCommands.java#L419

**Pull request merged**